src/ir/module-utils.h

```diff
@@ -262,33 +262,18 @@ template<typename T> inline void iterDefinedEvents(Module& wasm, T visitor) {
   }
 }
 
-// Helper class for analyzing the call graph.
-//
-// Provides hooks for running some initial calculation on each function (which
-// is done in parallel), writing to a FunctionInfo structure for each function.
-// Then you can call propagateBack() to propagate a property of interest to the
-// calling functions, transitively.
-//
-// For example, if some functions are known to call an import "foo", then you
-// can use this to find which functions call something that might eventually
-// reach foo, by initially marking the direct callers as "calling foo" and
-// propagating that backwards.
-template<typename T> struct CallGraphPropertyAnalysis {
+// Helper class for performing an operation on all the functions in the module,
+// in parallel, with an Info object for each one that can contain results of
+// some computation that the operation performs.
+template<typename T> struct ParallelFunctionAnalysis {
   Module& wasm;
 
-  // The basic information for each function about whom it calls and who is
-  // called by it.
-  struct FunctionInfo {
-    std::set<Function*> callsTo;
-    std::set<Function*> calledBy;
-  };
-
   typedef std::map<Function*, T> Map;
   Map map;
```

Comment:
Would it make sense to hide the map as an implementation detail and instead take some sort of `fold` function to be automatically applied to the results?

Reply:
Hmm, not sure what you mean? The map is user-observable, it's the output of the operation: a map with an info object per function, which were generated in parallel. What kind of `fold` do you have in mind to replace that?

Reply:
I mean have the user pass a function `combine (T, T) -> T` that must be commutative, then use that function to combine all the function results internally and return just a single T at the end, rather than producing a map of functions to Ts. That way each thread could combine its own results in parallel and the user wouldn't have to loop through the map and do the final combine themselves.

Reply:
Oh, well, Asyncify for example literally wants the map, as it does many later operations using that map.

Reply:
I guess this is more of a `reduce` than a `fold`

Reply:
Also, this would let `parallelFunctionAnalysis` be a function rather than a class.